In [0]:
from sklearn.datasets import fetch_20newsgroups

In [3]:
train=fetch_20newsgroups(subset='train',shuffle=True)

In [4]:
#Class Names
train.target_names

['alt.atheism',
 'comp.graphics',
 'comp.os.ms-windows.misc',
 'comp.sys.ibm.pc.hardware',
 'comp.sys.mac.hardware',
 'comp.windows.x',
 'misc.forsale',
 'rec.autos',
 'rec.motorcycles',
 'rec.sport.baseball',
 'rec.sport.hockey',
 'sci.crypt',
 'sci.electronics',
 'sci.med',
 'sci.space',
 'soc.religion.christian',
 'talk.politics.guns',
 'talk.politics.mideast',
 'talk.politics.misc',
 'talk.religion.misc']

In [5]:
print(train.data[2501])
print(train.target[2501])

From: johnh@macadam.mpce.mq.edu.au (John Haddy)
Subject: Re: HC11 blues -> no can find
Organization: Macquarie University
Lines: 38
Distribution: world
NNTP-Posting-Host: macadam.mpce.mq.edu.au

In article <pat.734048757@fegmania.wustl.edu>, pat@fegmania (Patrick Niemeyer) writes:
|> ree88132@zach.fit.edu (Keith Ledig) writes:
|> 
|> >was told by other people that it just takes a long time.  I heard
|> >rumors that the HC11 is very hard to get these days and is being
|> >replaced by the 711 series.  The sales person is trying to push
|> >the 711.  Can someone please tell me what is going on with these
|> >microcontrollers.  I can't seem to get a straight answer.
|> 
|> I too had trouble finding hc11's when I looked for them a while back.
|> I'd be interesting in hearing if anyone knows about their availability now...
|> 
|> Thanks,
|> Pat


The HC711 is an EPROM version of the HC11. Raw HC11 parts have factory set
ROM images, and as such are useless to the hobbyist. HC811 parts have
EE

In [6]:
print(train.target_names[12])

sci.electronics


In [0]:
from sklearn.feature_extraction.text import CountVectorizer

In [0]:
count_vect=CountVectorizer()

In [0]:
X_counts=count_vect.fit_transform(train.data)

In [10]:
X_counts.shape

(11314, 130107)

In [0]:
#Tf-idf
from sklearn.feature_extraction.text import TfidfTransformer


In [0]:
tfidf=TfidfTransformer()

In [0]:
X_tfidf=tfidf.fit_transform(X_counts)

In [14]:
X_tfidf.shape

(11314, 130107)

In [0]:
from sklearn.naive_bayes import MultinomialNB


In [0]:
clf=MultinomialNB()

In [17]:
clf.fit(X_tfidf,train.target)

MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)

In [0]:
from sklearn.pipeline import Pipeline

text_clf = Pipeline([('vect', CountVectorizer()), ('tfidf', TfidfTransformer()), ('clf', MultinomialNB())])

text_clf = text_clf.fit(train.data, train.target)


In [0]:
 TfidfTransformer()

In [0]:
import numpy as np
twenty_test = fetch_20newsgroups(subset='test', shuffle=True)
predicted = text_clf.predict(twenty_test.data)

In [20]:
np.mean(predicted == twenty_test.target)

0.7738980350504514

In [0]:
from sklearn.model_selection import GridSearchCV

In [0]:
parameters = {'vect__ngram_range': [(1, 1), (1, 2)],'tfidf__use_idf': (True, False),'clf__alpha': (1e-2, 1e-3)}


In [0]:
gs = GridSearchCV(text_clf, parameters, n_jobs=-1)


In [0]:
gs_fit = gs.fit(train.data,train.target)

In [30]:
gs_fit.best_score_

0.9067526957751458

In [31]:
gs_fit.best_params_

{'clf__alpha': 0.01, 'tfidf__use_idf': True, 'vect__ngram_range': (1, 2)}

## * Hence we can achieve 90.7% accuracy using above mentioned parameter values for built pipeline*